api/turing/cluster/knative_service_test.go

```diff
@@ -267,9 +268,92 @@ func TestBuildKnativeServiceConfig(t *testing.T) {
 	for name, data := range tests {
 		t.Run(name, func(t *testing.T) {
 			// Run test and validate
-			svc := data.serviceCfg.BuildKnativeServiceConfig()
-			err := tu.CompareObjects(*svc, data.expectedSpec)
+			svc, err := data.serviceCfg.BuildKnativeServiceConfig()
+			require.NoError(t, err)
+			err = tu.CompareObjects(*svc, data.expectedSpec)
 			assert.NoError(t, err)
 		})
 	}
 }
+
+func TestGetAutoscalingTarget(t *testing.T) {
+	tests := map[string]struct {
+		cfg            *KnativeService
+		expectedTarget string
+		expectedErr    string
+	}{
+		"concurrency": {
+			cfg: &KnativeService{
+				AutoscalingMetric: "concurrency",
+				AutoscalingTarget: "10",
+			},
+			expectedTarget: "10",
+		},
+		"rps": {
+			cfg: &KnativeService{
+				AutoscalingMetric: "rps",
+				AutoscalingTarget: "100",
+			},
+			expectedTarget: "100",
+		},
+		"cpu": {
+			cfg: &KnativeService{
+				AutoscalingMetric: "cpu",
+				AutoscalingTarget: "80",
+			},
+			expectedTarget: "80",
+		},
+		"memory | Mi": {
+			cfg: &KnativeService{
+				BaseService: &BaseService{
+					MemoryRequests: resource.MustParse("512Mi"),
+				},
+				AutoscalingMetric: "memory",
+				AutoscalingTarget: "70",
+			},
+			expectedTarget: "358", // (70/100) * (512 Mi)
+		},
+		"memory | Gi": {
+			cfg: &KnativeService{
+				BaseService: &BaseService{
+					MemoryRequests: resource.MustParse("2Gi"),
+				},
+				AutoscalingMetric: "memory",
+				AutoscalingTarget: "70",
+			},
+			expectedTarget: "1434", // (70/100) * (2Gi * 1024)
+		},
+		"memory | G": {
+			cfg: &KnativeService{
+				BaseService: &BaseService{
+					MemoryRequests: resource.MustParse("2G"),
+				},
+				AutoscalingMetric: "memory",
+				AutoscalingTarget: "70",
```

Comment:
Would a test case with `AutoscalingTarget` > 100 be useful?

Reply:
Good question. I considered it, but decided not to restrict the values. For example, we have the resource limit set to be a factor of the request. What if the user really only wants the autoscaling to kick in at 105% ? This is probably unlikely but I couldn't think of a strong reason to enforce this check.